metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java

```diff
@@ -109,14 +110,9 @@ public class StandardAuthorizerData {
     private final DefaultRule defaultRule;
 
     /**
-     * Contains all of the current ACLs sorted by (resource type, resource name).
+     * An immutable array of all the current ACLs sorted by (resource type, resource name).
      */
-    private final ConcurrentSkipListSet<StandardAcl> aclsByResource;
```

Comment:
I see we are replacing the skip-list set with a sorted array. Don't we need to guard against duplicates in the array?

If we used a TreeSet here, it would be closer to the current implementation and I think it should have linear time when copying from another TreeSet

Reply:
TreeSets are quite slow because they have poor memory locality. They also use a lot more memory. Ideally we could use something like a BTree, but Java doesn't have those, unfortunately...

Reply:
The downside of the current implementation is that updates are pretty expensive, right? Can we file a JIRA to investigate this further post 3.3?